wp/lib/bap_wp/src/output.ml

```diff
@@ -27,55 +28,116 @@ module FInterp = Model.FuncInterp
 module Env = Environment
 module Constr = Constraint
 module Comp = Compare
+module Quant = Z3.Quantifier
+module BV = Z3.BitVector
+module Sym = Z3.Symbol
+module AST = Z3.AST
 
 module VarMap = Var.Map
 
-type mem_model = {default : Constr.z3_expr ; model : (Constr.z3_expr * Constr.z3_expr) list}
+type mem_model = {
+  default : Constr.z3_expr;
+  model : (Constr.z3_expr * Constr.z3_expr) list
+}
+
+let equal_mem_model (m1 : mem_model) (m2 : mem_model) : bool =
+  let eqe (k1, v1) (k2, v2) = Expr.equal k1 k2 && Expr.equal v1 v2 in
+  let eqd = Expr.equal m1.default m2.default in
+  let eqm = List.equal eqe m1.model m2.model in
+  eqd && eqm
 
 let format_mem_model (fmt : Format.formatter) (mem_model : mem_model) : unit =
-  mem_model.model
-  |> List.sort ~compare:(fun (addr1, _) (addr2, _) ->
-      String.compare (Expr.to_string addr1) (Expr.to_string addr2))
-  |> List.iter ~f:(fun (key, data) ->
+  mem_model.model |> List.iter ~f:(fun (key, data) ->
```

Comment:
Is this list always going to be in order by address?

Reply:
As long as it's the one being produced by `extract_array`, then yes. I think that this function `format_mem_model` is not used anywhere outside of this module.